### PR TITLE
Fix the ability to disable buffering

### LINKED ISSUE #508: Unable to disable buffering once enabled
## Summary

Hi, I am migrating a web to net core. The line 68 enable response buffering if `HttpApplication` events are needed. This disables our cache policy in later code.

https://github.com/dotnet/systemweb-adapters/blob/c2607beb66e73455591084f8d3b73dd6eff2bc81/src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/SystemWebAdaptersExtensions.cs#L59-L81

## Motivation and goals

Allow user code to manage response buffering feature.

## Risks / unknowns

Can anyone provide some insights for this design?

## Repro

```c#
builder.Services.AddSystemWebAdapters()
    .AddWrappedAspNetCoreSession()
    .AddHttpApplication<MyApp>(options =>
    {
        options.RegisterModule<MyModule>("Module1");
        options.RegisterModule<MyModule>("Module2");
        options.RegisterModule<MyModule>("Module3");
.......
    });

var app = builder.Build();

// Since HttpApplication events are needed, here will add `HttpApplicationMiddleware`.
app.UseSystemWebAdapters();

app.Use((context, next) =>
{
    // This will not success because `HttpApplicationMiddleware` will enable buffering.
    if(!ShouldBufferReponse(context.Request.Path))
        context.Features.Get<IHttpResponseBodyFeature>().DisableBuffering();
    return next();
});
```
Here is where updated the response buffering.
https://github.com/dotnet/systemweb-adapters/blob/c2607beb66e73455591084f8d3b73dd6eff2bc81/src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpApplication/HttpApplicationMiddleware.cs#L18

### PR DESCRIPTION
**Pull request by twsouthwick**
If buffering is turned on (either because using HttpModules or the endpoint metadata), it was broken to call `ctx.Features.GetRequired<IHttpResponseBodyFeature>().DisableBuffering()`. This fixes that.

Fixes #508

### CODE REVIEW

**Comment by twsouthwick** [comment]
@joperezr can I get a review here?

**Comment by TaoziZ03** [comment]
There is a behavior mismatch.
If buffering is disabled and something is written. The framework `HttpApplication` will short circuit the request pipeline and jump to `RequestNotification.SendReponse` and raise `PreSendRequestHeaders` and `PreSendRequestContent` events.
In comparison systemweb-adapter will skip all the next stages and send the response.

**Comment by twsouthwick** [comment]
@TaoziZ03 I've moved the comment about the PreSendRequest* to a new issue and a fix is in #514

**Review by TaoziZ03 — APPROVED** [review]
Looks good, thanks!

**Comment by TaoziZ03** [comment]
A bug here is once the buffering disabled. The response will be sent to client after the first write. Please consider to raise `IHttpResponseEndFeature.EndAsync()` together.

Here is the Framework behavior. https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/System.Web/HttpWriter.cs#L1610

**Review by TaoziZ03 — CHANGES_REQUESTED** [review]
Please check this.
https://github.com/dotnet/systemweb-adapters/pull/509#issuecomment-2229894358

**Comment by twsouthwick** [comment]
When I run it on framework, those events are still raised:

```
BeginRequest
PreSendRequestHeaders
AuthenticateRequest
PostAuthenticateRequest
AuthorizeRequest
PostAuthorizeRequest
ResolveRequestCache
PostResolveRequestCache
MapRequestHandler
PostMapRequestHandler
AcquireRequestState
PostAcquireRequestState
PreRequestHandlerExecute
Hello world
PostRequestHandlerExecute
ReleaseRequestState
PostReleaseRequestState
UpdateRequestCache
PostUpdateRequestCache
LogRequest
PostLogRequest
EndRequest
```

Note the headers are sent early (and the content event isn't enabled in this pr). I've created #518 for this one, but in the future, can you create new issues for additional issues you find? This PR is focused on enabling the ability to disable buffering - by allowing that, we may find other issues, but let's get that in and deal with any follow up issues separately.

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs [hunk 1]
```diff
@@ -16,13 +16,15 @@ namespace Microsoft.AspNetCore.SystemWebAdapters.Features;
 [Experimental(Constants.ExperimentalFeatures.DiagnosticId)]
 public interface IHttpResponseBufferingFeature
 {
-    void EnableBuffering(int memoryThreshold, long? bufferLimit);
+    void EnableBuffering(int? memoryThreshold = default, long? bufferLimit = default);
```
Comment (joperezr) [inline src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs:19]: Do we really need the defaults and nullable changes here? Trying to see if we can avoid being binary breaking here.

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs [hunk 2]
```diff
@@ -16,13 +16,15 @@ namespace Microsoft.AspNetCore.SystemWebAdapters.Features;
...
 
     ValueTask FlushAsync();
 
     [AllowNull]
     Stream Filter { get; set; }
 
+    void DisableBuffering();
```
Comment (joperezr) [inline src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs:26]: Adding this member to the interface is of course also a breaking change. Do we expect this to cause any issues?

**Review by joperezr — APPROVED** [review]
Other than the two breaking change questions I have, changes look good to me.

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs (hunk 1 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs:19]: it makes it easier to just enable buffering when we want the default and the default value is currently in a different layer than I wanted to call it from.

This interface is intentionally marked as "Experimental" so we can break it as much as we need

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs (hunk 2 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/Features/IHttpResponseBufferingFeature.cs:26]: Same as above:  this interface is intentionally marked as "Experimental" so we can break it as much as we need

**Pull request merged**